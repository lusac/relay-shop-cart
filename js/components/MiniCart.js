import React from 'react'

export default class MiniCart extends React.Component {
  handleToggle() {
    if (!this.props.show)
      return 'hidden'
  }

  render() {
    return (
      <div className={'mini-cart ' + this.handleToggle()}>
        <div className="mini-cart__header">carrinho</div>
        <div className="mini-cart__section">
          <ul className="mini-cart__items">
            {Object.keys(this.props.items).map((id) => {
              return (
                <li key={id} className="mini-cart__item">
                  <img
                    className="mini-cart__item__img"
                    src={'/imgs/' + this.props.items[id].image}></img>
                  <span className="mini-cart__item__middle">
                    <h5 className="mini-cart__item__name">
                      {this.props.items[id].name}
                    </h5>
                    <small className="mini-cart__item__qty">
                      Quantidade: {this.props.items[id].qty}
                    </small>
                  </span>
                  <strong className="mini-cart__item__total">
                    R$ {this.props.items[id].qty * this.props.items[id].price}
                  </strong>
                </li>
              )
            })}
          </ul>
        </div>
        <div className="mini-cart__footer">
          total R$ {this.props.totalPrice}
        </div>
      </div>
    )
  }
}