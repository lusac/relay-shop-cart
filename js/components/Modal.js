import React from 'react'
import Modal from '@material-ui/core/Modal'
import Button from '@material-ui/core/Button'
import { withStyles } from '@material-ui/core/styles'
import Typography from '@material-ui/core/Typography'
import AddProductToCartMutation from '../mutations/AddProductToCartMutation'

function getModalStyle() {
  const top = 50
  const left = 50

  return {
    top: `${top}%`,
    left: `${left}%`,
    transform: `translate(-${top}%, -${left}%)`,
  }
}

const styles = theme => ({
  paper: {
    position: 'absolute',
    width: theme.spacing.unit * 50,
    backgroundColor: theme.palette.background.paper,
    boxShadow: theme.shadows[5],
    padding: theme.spacing.unit * 4,
  },
})

export class SimpleModal extends React.Component {
  _handleBuyProduct = () => {
    if (this.props.product.amount > 0) {
      AddProductToCartMutation.commit(
        this.props.relay.environment,
        this.props.product
      )
      return
    }

    alert('Produto Esgotado!')
  }

  render() {
    return (
      <div>
        {this.props.product && this.props.open &&
          <Modal
            open={this.props.open}
            onClose={this.props.handleClose}>
            <div style={getModalStyle()} className={this.props.classes.paper}>
              <img className="modal__img" src={'/imgs/' + this.props.product.image}></img>
              <Typography variant="title" className="modal__title">
                {this.props.product.name}
              </Typography>
              <Typography variant="subheading" className="modal__subtitle">
                Duis mollis, est non commodo luctus, nisi erat porttitor ligula.
              </Typography>
              <p className="modal__subtitle">
                Em estoque: {this.props.product.amount}
              </p>
              <strong className="modal__price">
                R$ {this.props.product.price.toFixed(2).replace(/\d(?=(\d{3})+\.)/g, '$&,')}
              </strong>
              <Button
                className="modal__button"
                disabled={this.props.product.amount > 0 ? false : true}
                onClick={this._handleBuyProduct}
                variant="contained"
                color="primary">
                Comprar
              </Button>
              <SimpleModalWrapped />
            </div>
          </Modal>
        }
      </div>
    )
  }
}

// We need an intermediary variable for handling the recursive nesting.
const SimpleModalWrapped = withStyles(styles)(SimpleModal)

export default SimpleModalWrapped