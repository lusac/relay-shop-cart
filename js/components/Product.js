import React from 'react';
import Button from '@material-ui/core/Button';
import {createFragmentContainer, graphql} from 'react-relay';

class Product extends React.Component {
  render() {
    return (
      <li className="product">
        <img
          className="product__image"
          src={'/imgs/' + this.props.product.image}></img>

        <strong className="product__name">
          {this.props.product.name} </strong>

        <span className="product__price">
          R$ {this.props.product.price.toFixed(2)} </span>

        <Button variant="contained" color="primary">
          Comprar </Button>
      </li>
    );
  }
}

export default createFragmentContainer(Product, {
  product: graphql`
    fragment Product_product on Product {
      complete
      id
      name
      image
      price
      amount
    }
  `,
  viewer: graphql`
    fragment Product_viewer on User {
      id
      totalCount
      completedCount
    }
  `,
});
