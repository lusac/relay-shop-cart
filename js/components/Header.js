import React from 'react'
import Badge from '@material-ui/core/Badge'
import AppBar from '@material-ui/core/AppBar'
import Toolbar from '@material-ui/core/Toolbar'
import IconButton from '@material-ui/core/IconButton'
import Typography from '@material-ui/core/Typography'
import ShoppingCartIcon from '@material-ui/icons/ShoppingCart'
import {createFragmentContainer, graphql} from 'react-relay'
import MiniCart from './MiniCart'


class Header extends React.Component {
  state = {
    showMiniCart: false
  }

  _toggleMiniCart = () => {
    this.setState({ showMiniCart: !this.state.showMiniCart })
  }

  _getCartTotalPrice = () => {
    let total = 0
    this.props.viewer.cart.edges.map((product) => {
      total += product.node.price
    })
    return `R$ ${total.toFixed(2)}`
  }

  _getFormatProductsDate = () => {
    // Melhorar estrutura do carrinho.
    // Já deveria vir como um dicionário.
    let products = {}
    this.props.viewer.cart.edges.map(item => {
      if (products[item.node.id]) {
        products[item.node.id].qty++
      } else {
        products[item.node.id] = {...item.node}
        products[item.node.id].qty = 1
      }
    })

    return products
  }

  render() {
    return (
      <AppBar position="static">
        <Toolbar>
          <Typography variant="title" color="inherit" noWrap>
            Minha Lojinha
          </Typography>

          <div>
            <IconButton
              color="inherit"
              onClick={this._toggleMiniCart.bind(this)}>
              <Badge badgeContent={this.props.viewer.cart.edges.length} color="secondary">
                <ShoppingCartIcon />
              </Badge>
            </IconButton>

            <MiniCart
              show={this.state.showMiniCart}
              items={this._getFormatProductsDate()}
              totalPrice={this._getCartTotalPrice()}
              environment={this.props.relay.environment} />
          </div>

          <small>{this._getCartTotalPrice()}</small>
        </Toolbar>
      </AppBar>
    )
  }
}

export default createFragmentContainer(Header, {
  viewer: graphql`
    fragment Header_viewer on User {
      cart(
        first: 2147483647 # max GraphQLInt
      ) @connection(key: "Header_cart") {
        edges {
          node {
            id
            price
            name
            image
            amount
          }
        }
      }
      id
      ...Product_viewer
    }
  `,
})
