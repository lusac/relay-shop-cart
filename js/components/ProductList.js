import React from 'react'
import {createFragmentContainer, graphql} from 'react-relay'

import Product from './Product'


class ProductList extends React.Component {
  renderProducts() {
    return this.props.products.map(product => (
      <Product key={product.id} product={product} viewer={this.props.viewer} />
    ))
  }

  render() {
    return (
      <section className="main">
        <ul className="product-list">{this.renderProducts()}</ul>
      </section>
    )
  }
}

export default createFragmentContainer(ProductList, {
  viewer: graphql`
    fragment ProductList_viewer on User {
      id
      ...Product_viewer
    }
  `
})
