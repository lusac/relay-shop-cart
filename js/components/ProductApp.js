import React from 'react';
import Header from './Header';
import ProductList from './ProductList';
import {createFragmentContainer, graphql} from 'react-relay';

class ProductApp extends React.Component {
  render() {
    return (
      <section className="productapp">
        <Header viewer={this.props.viewer} />
        <ProductList viewer={this.props.viewer} />
      </section>
    )
  }
}

export default createFragmentContainer(ProductApp, {
  viewer: graphql`
    fragment ProductApp_viewer on User {
      id
      ...Header_viewer
      ...ProductList_viewer
    }
  `
})
