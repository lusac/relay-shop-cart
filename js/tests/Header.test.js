import React from 'react'
import renderer from 'react-test-renderer'
import {Header} from '../components/Header'

describe('Header component', () => {
  describe('snapshot', () => {
    it('Header default', () => {
      const
        viewer = {
          cart: {
            edges: [{node: {id:1}},{node: {id:2}},{node: {id:3}}]
          }
        },
        relay = {enviroment: {}},
        renderedValue =  renderer.create(<Header viewer={viewer} relay={relay} />).toJSON()
      expect(renderedValue).toMatchSnapshot()
    })
  })
})