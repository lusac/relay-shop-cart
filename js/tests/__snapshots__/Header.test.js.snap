// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component snapshot Header default 1`] = `
<header
  className="MuiPaper-root-10 MuiPaper-elevation4-16 MuiAppBar-root-1 MuiAppBar-positionStatic-5 MuiAppBar-colorPrimary-8"
>
  <div
    className="MuiToolbar-root-37 MuiToolbar-regular-39 MuiToolbar-gutters-38 header__toolbar"
  >
    <h2
      className="MuiTypography-root-41 MuiTypography-title-47 MuiTypography-colorInherit-61 MuiTypography-noWrap-58"
    >
      Relay Shop
    </h2>
    <div>
      <small>
        R$ NaN
      </small>
      <button
        className="MuiButtonBase-root-73 MuiIconButton-root-67 MuiIconButton-colorInherit-68"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiIconButton-label-72"
        >
          <span
            className="MuiBadge-root-76"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root-81"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 18c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2zM1 2v2h2l3.6 7.59-1.35 2.45c-.16.28-.25.61-.25.96 0 1.1.9 2 2 2h12v-2H7.42c-.14 0-.25-.11-.25-.25l.03-.12.9-1.63h7.45c.75 0 1.41-.41 1.75-1.03l3.58-6.49c.08-.14.12-.31.12-.48 0-.55-.45-1-1-1H5.21l-.94-2H1zm16 16c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
            <span
              className="MuiBadge-badge-77 MuiBadge-colorSecondary-79"
            >
              3
            </span>
          </span>
        </span>
        <span
          className="MuiTouchRipple-root-90"
        />
      </button>
      <div
        className="mini-cart--hidden"
      >
        <div
          className="mini-cart__overlay"
          onClick={[Function]}
        />
        <div
          className="mini-cart"
        >
          <div
            className="mini-cart__header"
          >
            carrinho
          </div>
          <div
            className="mini-cart__section"
          >
            <ul
              className="mini-cart__items"
            >
              <li
                className="mini-cart__item"
              >
                <span
                  className="mini-cart__item__remove"
                  onClick={[Function]}
                >
                  X
                </span>
                <img
                  className="mini-cart__item__img"
                  src="/imgs/undefined"
                />
                <span
                  className="mini-cart__item__middle"
                >
                  <h5
                    className="mini-cart__item__name"
                  />
                  <small
                    className="mini-cart__item__qty"
                  >
                    Quantidade: 
                    1
                  </small>
                  <span
                    className="mini-cart__change-qty"
                  >
                    <span
                      onClick={[Function]}
                    >
                      aumentar
                    </span>
                    <span
                      onClick={[Function]}
                    >
                      diminuir
                    </span>
                  </span>
                </span>
                <strong
                  className="mini-cart__item__total"
                >
                  R$ 
                  NaN
                </strong>
              </li>
              <li
                className="mini-cart__item"
              >
                <span
                  className="mini-cart__item__remove"
                  onClick={[Function]}
                >
                  X
                </span>
                <img
                  className="mini-cart__item__img"
                  src="/imgs/undefined"
                />
                <span
                  className="mini-cart__item__middle"
                >
                  <h5
                    className="mini-cart__item__name"
                  />
                  <small
                    className="mini-cart__item__qty"
                  >
                    Quantidade: 
                    1
                  </small>
                  <span
                    className="mini-cart__change-qty"
                  >
                    <span
                      onClick={[Function]}
                    >
                      aumentar
                    </span>
                    <span
                      onClick={[Function]}
                    >
                      diminuir
                    </span>
                  </span>
                </span>
                <strong
                  className="mini-cart__item__total"
                >
                  R$ 
                  NaN
                </strong>
              </li>
              <li
                className="mini-cart__item"
              >
                <span
                  className="mini-cart__item__remove"
                  onClick={[Function]}
                >
                  X
                </span>
                <img
                  className="mini-cart__item__img"
                  src="/imgs/undefined"
                />
                <span
                  className="mini-cart__item__middle"
                >
                  <h5
                    className="mini-cart__item__name"
                  />
                  <small
                    className="mini-cart__item__qty"
                  >
                    Quantidade: 
                    1
                  </small>
                  <span
                    className="mini-cart__change-qty"
                  >
                    <span
                      onClick={[Function]}
                    >
                      aumentar
                    </span>
                    <span
                      onClick={[Function]}
                    >
                      diminuir
                    </span>
                  </span>
                </span>
                <strong
                  className="mini-cart__item__total"
                >
                  R$ 
                  NaN
                </strong>
              </li>
            </ul>
          </div>
          <div
            className="mini-cart__footer"
          >
            total R$ 
            R$ NaN
          </div>
        </div>
      </div>
    </div>
  </div>
</header>
`;
